File: tetris.go
// tetris.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"os/exec"
	"runtime"
	"strconv"
	"strings"
	"time"
)

const (
	WIDTH  = 10
	HEIGHT = 20
)

type Piece struct {
	Shape [][]int
	X     int
	Y     int
	Name  string
	Color string
}

type Tetris struct {
	Board          [][]int
	Score          int
	Level          int
	LinesCleared   int
	GameOver       bool
	Paused         bool
	CurrentPiece   *Piece
	NextPiece      *Piece
	FallInterval   float64
	FallTime       float64
	HighScore      int
	Running        bool
	InputChan      chan string
}

func NewTetris() *Tetris {
	t := &Tetris{
		Board:        make([][]int, HEIGHT),
		Score:        0,
		Level:        1,
		LinesCleared: 0,
		GameOver:     false,
		Paused:       false,
		FallInterval: 1.0,
		FallTime:     0.0,
		HighScore:    0,
		Running:      true,
		InputChan:    make(chan string, 10),
	}
	for i := range t.Board {
		t.Board[i] = make([]int, WIDTH)
	}
	t.HighScore = t.loadHighScore()
	t.NextPiece = t.newPiece()
	t.spawnPiece()
	return t
}

func (t *Tetris) loadHighScore() int {
	data, err := os.ReadFile("tetris_score.json")
	if err != nil {
		return 0
	}
	var score map[string]int
	if err := json.Unmarshal(data, &score); err != nil {
		return 0
	}
	return score["high_score"]
}

func (t *Tetris) saveHighScore() {
	data, _ := json.Marshal(map[string]int{"high_score": t.HighScore})
	os.WriteFile("tetris_score.json", data, 0644)
}

func (t *Tetris) newPiece() *Piece {
	shapes := map[string][][]int{
		"I": {{1, 1, 1, 1}},
		"O": {{1, 1}, {1, 1}},
		"T": {{0, 1, 0}, {1, 1, 1}},
		"S": {{0, 1, 1}, {1, 1, 0}},
		"Z": {{1, 1, 0}, {0, 1, 1}},
		"J": {{1, 0, 0}, {1, 1, 1}},
		"L": {{0, 0, 1}, {1, 1, 1}},
	}
	names := []string{"I", "O", "T", "S", "Z", "J", "L"}
	name := names[rand.Intn(len(names))]
	shape := shapes[name]
	colors := map[string]string{
		"I": "cyan", "O": "yellow", "T": "magenta",
		"S": "green", "Z": "red", "J": "blue", "L": "white",
	}
	x := (WIDTH - len(shape[0])) / 2
	return &Piece{
		Shape: shape,
		X:     x,
		Y:     0,
		Name:  name,
		Color: colors[name],
	}
}

func (t *Tetris) spawnPiece() {
	if t.NextPiece != nil {
		t.CurrentPiece = t.NextPiece
	} else {
		t.CurrentPiece = t.newPiece()
	}
	t.NextPiece = t.newPiece()
	if !t.validPosition(t.CurrentPiece.Shape, t.CurrentPiece.X, t.CurrentPiece.Y) {
		t.GameOver = true
	}
}

func (t *Tetris) validPosition(shape [][]int, x, y int) bool {
	for rowIdx, row := range shape {
		for colIdx, cell := range row {
			if cell == 1 {
				newX := x + colIdx
				newY := y + rowIdx
				if newX < 0 || newX >= WIDTH || newY >= HEIGHT {
					return false
				}
				if newY >= 0 && t.Board[newY][newX] != 0 {
					return false
				}
			}
		}
	}
	return true
}

func (t *Tetris) lockPiece() {
	if t.CurrentPiece == nil {
		return
	}
	for rowIdx, row := range t.CurrentPiece.Shape {
		for colIdx, cell := range row {
			if cell == 1 {
				boardY := t.CurrentPiece.Y + rowIdx
				boardX := t.CurrentPiece.X + colIdx
				if boardY < 0 {
					t.GameOver = true
					return
				}
				t.Board[boardY][boardX] = 1 // цвет будет отображаться
			}
		}
	}
	t.clearLines()
	t.spawnPiece()
}

func (t *Tetris) clearLines() {
	linesCleared := 0
	newBoard := make([][]int, 0, HEIGHT)
	for _, row := range t.Board {
		hasZero := false
		for _, cell := range row {
			if cell == 0 {
				hasZero = true
				break
			}
		}
		if hasZero {
			newBoard = append(newBoard, row)
		} else {
			linesCleared++
		}
	}
	for i := 0; i < linesCleared; i++ {
		newBoard = append([][]int{make([]int, WIDTH)}, newBoard...)
	}
	t.Board = newBoard
	if linesCleared > 0 {
		t.LinesCleared += linesCleared
		scores := []int{0, 100, 300, 500, 800}
		addScore := scores[linesCleared]
		if linesCleared > 4 {
			addScore = 800
		}
		t.Score += addScore * t.Level
		t.Level = t.LinesCleared/10 + 1
		t.FallInterval = 1.0 - float64(t.Level-1)*0.07
		if t.FallInterval < 0.1 {
			t.FallInterval = 0.1
		}
		if t.Score > t.HighScore {
			t.HighScore = t.Score
			t.saveHighScore()
		}
	}
}

func (t *Tetris) rotateShape(shape [][]int) [][]int {
	rows := len(shape)
	cols := len(shape[0])
	rotated := make([][]int, cols)
	for i := range rotated {
		rotated[i] = make([]int, rows)
	}
	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			rotated[j][rows-1-i] = shape[i][j]
		}
	}
	return rotated
}

func (t *Tetris) movePiece(dx, dy int) bool {
	if t.GameOver || t.Paused || t.CurrentPiece == nil {
		return false
	}
	newX := t.CurrentPiece.X + dx
	newY := t.CurrentPiece.Y + dy
	if t.validPosition(t.CurrentPiece.Shape, newX, newY) {
		t.CurrentPiece.X = newX
		t.CurrentPiece.Y = newY
		return true
	}
	return false
}

func (t *Tetris) rotatePiece() {
	if t.GameOver || t.Paused || t.CurrentPiece == nil {
		return
	}
	rotated := t.rotateShape(t.CurrentPiece.Shape)
	if t.validPosition(rotated, t.CurrentPiece.X, t.CurrentPiece.Y) {
		t.CurrentPiece.Shape = rotated
	}
}

func (t *Tetris) hardDrop() {
	if t.GameOver || t.Paused || t.CurrentPiece == nil {
		return
	}
	for t.movePiece(0, 1) {
	}
	t.lockPiece()
}

func (t *Tetris) update(dt float64) {
	if t.GameOver || t.Paused {
		return
	}
	t.FallTime += dt
	if t.FallTime >= t.FallInterval {
		t.FallTime = 0
		if !t.movePiece(0, 1) {
			t.lockPiece()
		}
	}
}

func (t *Tetris) draw() {
	clearScreen()
	fmt.Printf("Тетрис (с физикой)   Счёт: %d   Уровень: %d   Рекорд: %d\n", t.Score, t.Level, t.HighScore)
	fmt.Println("┌" + strings.Repeat("─", WIDTH*2+1) + "┐")
	for y := 0; y < HEIGHT; y++ {
		fmt.Print("│")
		for x := 0; x < WIDTH; x++ {
			cell := t.Board[y][x]
			drawn := false
			if t.CurrentPiece != nil && !t.GameOver {
				for rowIdx, row := range t.CurrentPiece.Shape {
					for colIdx, val := range row {
						if val == 1 && t.CurrentPiece.Y+rowIdx == y && t.CurrentPiece.X+colIdx == x {
							color := colorCode(t.CurrentPiece.Color)
							fmt.Print(color + "██" + resetColor())
							drawn = true
							break
						}
					}
					if drawn {
						break
					}
				}
			}
			if !drawn {
				if cell != 0 {
					fmt.Print("\033[37m██\033[0m")
				} else {
					fmt.Print("  ")
				}
			}
		}
		fmt.Println("│")
	}
	fmt.Println("└" + strings.Repeat("─", WIDTH*2+1) + "┘")
	if t.NextPiece != nil {
		fmt.Println("Следующая:")
		for _, row := range t.NextPiece.Shape {
			fmt.Print("  ")
			for _, cell := range row {
				if cell == 1 {
					color := colorCode(t.NextPiece.Color)
					fmt.Print(color + "██" + resetColor())
				} else {
					fmt.Print("  ")
				}
			}
			fmt.Println()
		}
	}
	if t.GameOver {
		fmt.Println("\033[31mИГРА ОКОНЧЕНА! Нажмите R для рестарта\033[0m")
	} else if t.Paused {
		fmt.Println("\033[33mПАУЗА\033[0m")
	}
}

func colorCode(name string) string {
	switch name {
	case "cyan": return "\033[36m"
	case "yellow": return "\033[33m"
	case "magenta": return "\033[35m"
	case "green": return "\033[32m"
	case "red": return "\033[31m"
	case "blue": return "\033[34m"
	default: return "\033[37m"
	}
}

func resetColor() string { return "\033[0m" }

func clearScreen() {
	cmd := exec.Command("clear")
	if runtime.GOOS == "windows" {
		cmd = exec.Command("cmd", "/c", "cls")
	}
	cmd.Stdout = os.Stdout
	cmd.Run()
}

func (t *Tetris) inputLoop() {
	scanner := bufio.NewScanner(os.Stdin)
	for t.Running {
		if scanner.Scan() {
			line := scanner.Text()
			t.InputChan <- line
		}
	}
}

func (t *Tetris) run() {
	fmt.Println("Тетрис (с физикой)")
	fmt.Println("Управление: ← →, ↑ вращение, ↓ ускорение, Space - мгновенное падение, P - пауза, Q - выход")
	fmt.Println("Нажмите Enter для начала...")
	fmt.Scanln()

	go t.inputLoop()

	lastTime := time.Now()
	for t.Running {
		now := time.Now()
		dt := now.Sub(lastTime).Seconds()
		lastTime = now

		// Обработка ввода
		select {
		case input := <-t.InputChan:
			switch input {
			case "q", "Q":
				t.Running = false
			case "p", "P":
				t.Paused = !t.Paused
			case "r", "R":
				if t.GameOver {
					*t = *NewTetris()
					t.Running = true
				}
			default:
				if !t.Paused && !t.GameOver {
					switch input {
					case "left", "a":
						t.movePiece(-1, 0)
					case "right", "d":
						t.movePiece(1, 0)
					case "up", "w":
						t.rotatePiece()
					case "down", "s":
						t.movePiece(0, 1)
					case " ":
						t.hardDrop()
					}
				}
			}
		default:
		}

		t.update(dt)
		t.draw()
		time.Sleep(30 * time.Millisecond)
	}
}

func main() {
	rand.Seed(time.Now().UnixNano())
	game := NewTetris()
	game.run()
}

File: tetris.rs
// tetris.rs
use std::collections::HashMap;
use std::io::{self, Write, Read};
use std::time::{Duration, Instant};
use rand::Rng;
use termion::raw::IntoRawMode;
use termion::input::TermRead;
use colored::*;

const WIDTH: usize = 10;
const HEIGHT: usize = 20;

type Shape = Vec<Vec<u8>>;

struct Piece {
    shape: Shape,
    x: usize,
    y: usize,
    name: String,
    color: String,
}

impl Piece {
    fn rotate(&self) -> Shape {
        let rows = self.shape.len();
        let cols = self.shape[0].len();
        let mut rotated = vec![vec![0; rows]; cols];
        for i in 0..rows {
            for j in 0..cols {
                rotated[j][rows - 1 - i] = self.shape[i][j];
            }
        }
        rotated
    }
}

struct Tetris {
    board: Vec<Vec<u8>>,
    score: u32,
    level: u32,
    lines_cleared: u32,
    game_over: bool,
    paused: bool,
    current_piece: Option<Piece>,
    next_piece: Option<Piece>,
    fall_interval: f64,
    fall_time: f64,
    high_score: u32,
    running: bool,
}

impl Tetris {
    fn new() -> Self {
        let mut tetris = Tetris {
            board: vec![vec![0; WIDTH]; HEIGHT],
            score: 0,
            level: 1,
            lines_cleared: 0,
            game_over: false,
            paused: false,
            current_piece: None,
            next_piece: None,
            fall_interval: 1.0,
            fall_time: 0.0,
            high_score: 0,
            running: true,
        };
        tetris.high_score = tetris.load_high_score();
        tetris.next_piece = Some(tetris.new_piece());
        tetris.spawn_piece();
        tetris
    }

    fn load_high_score(&self) -> u32 {
        if let Ok(data) = std::fs::read_to_string("tetris_score.json") {
            if let Ok(json) = serde_json::from_str::<serde_json::Value>(&data) {
                if let Some(score) = json.get("high_score").and_then(|v| v.as_u64()) {
                    return score as u32;
                }
            }
        }
        0
    }

    fn save_high_score(&self) {
        let json = serde_json::json!({ "high_score": self.high_score });
        let _ = std::fs::write("tetris_score.json", serde_json::to_string_pretty(&json).unwrap());
    }

    fn new_piece(&self) -> Piece {
        let shapes: HashMap<&str, Shape> = [
            ("I", vec![vec![1,1,1,1]]),
            ("O", vec![vec![1,1], vec![1,1]]),
            ("T", vec![vec![0,1,0], vec![1,1,1]]),
            ("S", vec![vec![0,1,1], vec![1,1,0]]),
            ("Z", vec![vec![1,1,0], vec![0,1,1]]),
            ("J", vec![vec![1,0,0], vec![1,1,1]]),
            ("L", vec![vec![0,0,1], vec![1,1,1]]),
        ].iter().cloned().collect();
        let names = ["I","O","T","S","Z","J","L"];
        let name = names[rand::thread_rng().gen_range(0..names.len())];
        let shape = shapes[name].unwrap().clone();
        let colors = HashMap::from([
            ("I", "cyan"), ("O", "yellow"), ("T", "magenta"),
            ("S", "green"), ("Z", "red"), ("J", "blue"), ("L", "white")
        ]);
        let color = colors[name].unwrap().to_string();
        let x = (WIDTH - shape[0].len()) / 2;
        let y = 0;
        Piece { shape, x, y, name: name.to_string(), color }
    }

    fn spawn_piece(&mut self) {
        if let Some(next) = self.next_piece.take() {
            self.current_piece = Some(next);
        } else {
            self.current_piece = Some(self.new_piece());
        }
        self.next_piece = Some(self.new_piece());
        if let Some(piece) = &self.current_piece {
            if !self.valid_position(&piece.shape, piece.x, piece.y) {
                self.game_over = true;
            }
        }
    }

    fn valid_position(&self, shape: &Shape, offset_x: usize, offset_y: usize) -> bool {
        for (y, row) in shape.iter().enumerate() {
            for (x, &cell) in row.iter().enumerate() {
                if cell == 1 {
                    let new_x = offset_x + x;
                    let new_y = offset_y + y;
                    if new_x >= WIDTH || new_y >= HEIGHT {
                        return false;
                    }
                    if new_y < HEIGHT && self.board[new_y][new_x] != 0 {
                        return false;
                    }
                }
            }
        }
        true
    }

    fn lock_piece(&mut self) {
        if let Some(piece) = &self.current_piece {
            for (y, row) in piece.shape.iter().enumerate() {
                for (x, &cell) in row.iter().enumerate() {
                    if cell == 1 {
                        let board_y = piece.y + y;
                        let board_x = piece.x + x;
                        if board_y < HEIGHT {
                            self.board[board_y][board_x] = 1; // цвет будет обрабатываться при отрисовке
                        } else {
                            self.game_over = true;
                            return;
                        }
                    }
                }
            }
            self.clear_lines();
            self.spawn_piece();
        }
    }

    fn clear_lines(&mut self) {
        let mut lines_cleared = 0;
        let mut new_board = Vec::new();
        for row in &self.board {
            if row.iter().any(|&c| c == 0) {
                new_board.push(row.clone());
            } else {
                lines_cleared += 1;
            }
        }
        for _ in 0..lines_cleared {
            new_board.insert(0, vec![0; WIDTH]);
        }
        self.board = new_board;
        if lines_cleared > 0 {
            self.lines_cleared += lines_cleared as u32;
            let scores = [0, 100, 300, 500, 800];
            let add_score = scores[lines_cleared.min(4)] * self.level;
            self.score += add_score;
            self.level = self.lines_cleared / 10 + 1;
            self.fall_interval = (1.0 - (self.level - 1) as f64 * 0.07).max(0.1);
            if self.score > self.high_score {
                self.high_score = self.score;
                self.save_high_score();
            }
        }
    }

    fn move_piece(&mut self, dx: i32, dy: i32) -> bool {
        if self.game_over || self.paused { return false; }
        if let Some(piece) = &mut self.current_piece {
            let new_x = piece.x as i32 + dx;
            let new_y = piece.y as i32 + dy;
            if new_x >= 0 && new_y >= 0 {
                if self.valid_position(&piece.shape, new_x as usize, new_y as usize) {
                    piece.x = new_x as usize;
                    piece.y = new_y as usize;
                    return true;
                }
            }
        }
        false
    }

    fn rotate_piece(&mut self) {
        if self.game_over || self.paused { return; }
        if let Some(piece) = &mut self.current_piece {
            let rotated = piece.rotate();
            if self.valid_position(&rotated, piece.x, piece.y) {
                piece.shape = rotated;
            }
        }
    }

    fn hard_drop(&mut self) {
        if self.game_over || self.paused { return; }
        while self.move_piece(0, 1) {}
        self.lock_piece();
    }

    fn update(&mut self, dt: f64) {
        if self.game_over || self.paused { return; }
        self.fall_time += dt;
        if self.fall_time >= self.fall_interval {
            self.fall_time = 0.0;
            if !self.move_piece(0, 1) {
                self.lock_piece();
            }
        }
    }

    fn draw(&self, stdout: &mut termion::raw::RawTerminal<std::io::Stdout>) {
        write!(stdout, "{}", termion::clear::All).unwrap();
        write!(stdout, "{}", termion::cursor::Goto(1, 1)).unwrap();
        println!("Тетрис (с физикой)   Счёт: {}   Уровень: {}   Рекорд: {}", self.score, self.level, self.high_score);
        println!("┌{}┐", "─".repeat(WIDTH * 2 + 1));
        for y in 0..HEIGHT {
            print!("│");
            for x in 0..WIDTH {
                let cell = self.board[y][x];
                let mut drawn = false;
                if let Some(piece) = &self.current_piece {
                    if !self.game_over {
                        for (py, row) in piece.shape.iter().enumerate() {
                            for (px, &val) in row.iter().enumerate() {
                                if val == 1 && piece.y + py == y && piece.x + px == x {
                                    let color = match piece.color.as_str() {
                                        "cyan" => "█".cyan(),
                                        "yellow" => "█".yellow(),
                                        "magenta" => "█".magenta(),
                                        "green" => "█".green(),
                                        "red" => "█".red(),
                                        "blue" => "█".blue(),
                                        _ => "█".white(),
                                    };
                                    print!("{}", color);
                                    drawn = true;
                                    break;
                                }
                            }
                            if drawn { break; }
                        }
                    }
                }
                if !drawn {
                    if cell != 0 {
                        print!("{}", "█".white());
                    } else {
                        print!("  ");
                    }
                }
            }
            println!("│");
        }
        println!("└{}┘", "─".repeat(WIDTH * 2 + 1));
        // следующая фигура
        if let Some(next) = &self.next_piece {
            println!("Следующая:");
            for row in &next.shape {
                print!("  ");
                for &cell in row {
                    if cell == 1 {
                        let color = match next.color.as_str() {
                            "cyan" => "██".cyan(),
                            "yellow" => "██".yellow(),
                            "magenta" => "██".magenta(),
                            "green" => "██".green(),
                            "red" => "██".red(),
                            "blue" => "██".blue(),
                            _ => "██".white(),
                        };
                        print!("{}", color);
                    } else {
                        print!("  ");
                    }
                }
                println!();
            }
        }
        if self.game_over {
            println!("{}", "ИГРА ОКОНЧЕНА! Нажмите R для рестарта".red());
        } else if self.paused {
            println!("{}", "ПАУЗА".yellow());
        }
        stdout.flush().unwrap();
    }

    fn run(&mut self) {
        let mut stdout = io::stdout().into_raw_mode().unwrap();
        let stdin = io::stdin();
        let mut keys = stdin.keys();

        write!(stdout, "{}", termion::clear::All).unwrap();
        write!(stdout, "{}", termion::cursor::Goto(1, 1)).unwrap();
        println!("Тетрис (с физикой)");
        println!("Управление: ← →, ↑ вращение, ↓ ускорение, Space - мгновенное падение, P - пауза, Q - выход");
        println!("Нажмите любую клавишу для начала...");
        stdout.flush().unwrap();
        keys.next(); // ждём клавишу

        let mut last_update = Instant::now();
        while self.running {
            let now = Instant::now();
            let dt = now - last_update;
            last_update = now;

            // Ввод
            if let Some(Ok(key)) = keys.next() {
                match key {
                    termion::event::Key::Char('q') | termion::event::Key::Char('Q') => self.running = false,
                    termion::event::Key::Char('p') | termion::event::Key::Char('P') => self.paused = !self.paused,
                    termion::event::Key::Char('r') | termion::event::Key::Char('R') if self.game_over => {
                        *self = Tetris::new();
                        self.running = true;
                    }
                    termion::event::Key::Left | termion::event::Key::Char('a') => { self.move_piece(-1, 0); }
                    termion::event::Key::Right | termion::event::Key::Char('d') => { self.move_piece(1, 0); }
                    termion::event::Key::Up | termion::event::Key::Char('w') => { self.rotate_piece(); }
                    termion::event::Key::Down | termion::event::Key::Char('s') => { self.move_piece(0, 1); }
                    termion::event::Key::Char(' ') => { self.hard_drop(); }
                    _ => {}
                }
            }

            self.update(dt.as_secs_f64());
            self.draw(&mut stdout);
            std::thread::sleep(std::time::Duration::from_millis(30));
        }
    }
}

fn main() {
    let mut game = Tetris::new();
    game.run();
}
